drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-backend/src/main/java/org/drools/workbench/screens/scenariosimulation/backend/server/DMNTypeServiceImpl.java

```diff
@@ -122,14 +124,18 @@ protected FactModelTree createTopLevelFactModelTree(String factName, DMNType typ
      * @return
      * @throws WrongDMNTypeException if <code>DMNType.isCollection()</code> != <code>true</code>
      */
-    protected FactModelTree createFactModelTreeForCollection(Map<String, List<String>> genericTypeInfoMap, String factName, DMNType type, SortedMap<String, FactModelTree> hiddenFacts, FactModelTree.Type fmType) throws WrongDMNTypeException {
+    protected FactModelTree createFactModelTreeForCollection(Map<String, List<String>> genericTypeInfoMap, String factName, DMNType type, SortedMap<String, FactModelTree> hiddenFacts, FactModelTree.Type fmType, List<String> visitedTypes) throws WrongDMNTypeException {
         if (!type.isCollection() && !isToBeManagedAsCollection((BaseDMNTypeImpl) type)) {
             throw new WrongDMNTypeException();
         }
-        populateGeneric(genericTypeInfoMap, "value", type.getName());
+        String typeName = type.getName();
+        populateGeneric(genericTypeInfoMap, "value", typeName);
         FactModelTree toReturn = createFactModelTreeForSimple(genericTypeInfoMap, factName, List.class.getCanonicalName(), fmType);
-        FactModelTree genericFactModelTree = createFactModelTreeForGenericType(new HashMap<>(), type.getName(), type.getName(), type.getName(), type, hiddenFacts, fmType);
-        hiddenFacts.put(type.getName(), genericFactModelTree);
+        if (!hiddenFacts.containsKey(typeName) && !visitedTypes.contains(typeName)) {
```

Comment:
@danielezonca 
I'm not reading the whole code, and I remember there was a reason for that. But seeing this _if_ now, it seems that _visited_ types is a little bit redundant.

Reply:
They are different: `hiddenFacts` contains already visited types (visited entirely) so it avoid to revisit them while `visitedTypes` contains the types you are analyzing now to avoid infinite recursions.
I.e. in a structure like `Person.parent` the execution will reach this point with the root `Person` type not completed but code has to consider it while analyzing `parent` to avoid the recursion